CUDADataFormats/Common/interface/HostProduct.h

```diff
@@ -0,0 +1,40 @@
+#ifndef CUDADataFormatsCommonHostProduct_H
+#define CUDADataFormatsCommonHostProduct_H
+
+#include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
+
+// a heterogeneous unique pointer...
+template<typename T>
+class HostProduct {
+public:
+
+
+  HostProduct() = default; // make root happy
+  ~HostProduct() = default;
+  HostProduct(HostProduct&&) = default;
+  HostProduct& operator=(HostProduct&&) = default;
+
+  explicit HostProduct(cudautils::host::unique_ptr<T> && p) : hm_ptr(std::move(p)) {}
+  explicit HostProduct(std::unique_ptr<T> && p) : std_ptr(std::move(p)) {}
+
+
+  T const * get() const {
+    return hm_ptr ? hm_ptr.get() : std_ptr.get();
+  }
+  
+  T const & operator*() const {
+    return *get();
+  }
+
+  T const * operator->() const {
+    return get();
+  }
+  
+
+private:
+  cudautils::host::unique_ptr<T> hm_ptr;
+  std::unique_ptr<T> std_ptr;
```

Comment:
given that most probably we should move to malloc/free instead of new/delete for CPU allocation
(we use malloc/free on GPU/HOST and assume constructor and destructor are not required)
I think we should anyhow review this and the other multiple-ptr class (and the associated Traits)

Reply:
I'm also thinking that it's probably best to include all this now, and then review, think further, and polish.